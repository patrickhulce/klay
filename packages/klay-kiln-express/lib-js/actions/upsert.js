const creatifyModel = require('../shared/creatifyModel')
const findDbModel = require('../shared/findDbModel')

module.exports = {
  options: {},
  bodyModel(modelDef, options) {
    return creatifyModel(modelDef.model, options)
  },
  handler(modelDef, options, extOptions, dependencies) {
    const dbModel = findDbModel(modelDef.name, options, dependencies)
    return function (req, res, next) {
      const payload = req.validated.body
      res.promise = dbModel.upsert(payload)
      next()
    }
  },
}
