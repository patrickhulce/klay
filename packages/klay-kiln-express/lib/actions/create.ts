import {NextFunction, Request, Response} from 'express'
import {IModel} from 'klay-core'
import {IDatabaseExecutor} from 'klay-db'
import {IKilnModel} from 'klay-kiln'
import {creatifyModel} from '../helpers/transform-model'
import {ActionType, IAction, IActionOptions, IAnontatedHandler} from '../typedefs'
import {defaultAction} from './action'

export const createAction: IAction = {
  ...defaultAction,
  type: ActionType.Create,
  bodyModel(kilnModel: IKilnModel, options: IActionOptions): IModel {
    return creatifyModel(kilnModel.model)
  },
  handler(
    model: IKilnModel,
    options: IActionOptions,
    executor: IDatabaseExecutor,
  ): IAnontatedHandler {
    return function(req: Request, res: Response, next: NextFunction): void {
      const payload = req.validated!.body
      res.promise = executor.create(payload)
      next()
    }
  },
}
