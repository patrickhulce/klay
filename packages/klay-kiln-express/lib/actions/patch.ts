import {NextFunction, Request, Response} from 'express'
import {IModel} from 'klay-core'
import {IDatabaseExecutor} from 'klay-db'
import {IKilnModel} from 'klay-kiln'

import {paramifyModel, updateifyModel} from '../helpers/transform-model'
import {
  ActionType,
  AuthCriteriaValue,
  GetCriteriaValues,
  IAction,
  IActionOptions,
  IAnontatedHandler,
} from '../typedefs'

import {defaultAction} from './action'

export const patchAction: IAction = {
  ...defaultAction,
  type: ActionType.Patch,
  defaultOptions: {
    byId: true, // MUST be set to true
    idParamName: undefined,
    patchProperties: undefined,
  },
  getAffectedCriteriaValues(model: IKilnModel, options: IActionOptions): GetCriteriaValues {
    return function(req: Request, property: string): AuthCriteriaValue[] {
      const incoming = {...req.actionTarget, ...req.validated!.body}
      return [incoming[property], req.actionTarget[property]]
    }
  },
  paramsModel(kilnModel: IKilnModel, options: IActionOptions): IModel {
    return paramifyModel(kilnModel.model, options)
  },
  bodyModel(kilnModel: IKilnModel, options: IActionOptions): IModel {
    if (!options.patchProperties || !options.patchProperties.length)
      throw new Error('Cannot patch without properties')

    // TODO: assert that properties actually exist on model
    const updateModel = updateifyModel(kilnModel.model)
    return updateModel.pick(options.patchProperties)
  },
  handler(
    model: IKilnModel,
    options: IActionOptions,
    executor: IDatabaseExecutor,
  ): IAnontatedHandler {
    return function(req: Request, res: Response, next: NextFunction): void {
      const payload = req.validated!.body
      const updated = {...req.actionTarget, ...payload}
      res.promise = executor.update(updated)
      next()
    }
  },
}
