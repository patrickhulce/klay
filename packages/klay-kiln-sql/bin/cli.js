#!/usr/bin/env node
/* eslint-disable no-console */

const fs = require('fs')
const path = require('path')
const Umzug = require('umzug')
const yargs = require('yargs')
const colors = require('colors')
const Sequelize = require('sequelize')

const getConnection = require('../dist/sequelize/connection').getConnection
const createNewMigrationFile = require('../dist/gen/migration').createNewMigrationFile

const options = yargs
  .command('migrate', 'run all pending migrations', {url: {required: true}})
  .command('migrate:revert', 'undo last migration', {url: {required: true}})
  .command('migration:bootstrap', 'create a migration file', {
    kilnFile: {alias: 'k', required: true, describe: 'path to file exporting kiln instance'},
  })
  .command('migration:incremental', 'create an incremental migration file', {
    url: {required: true, describe: 'SQL connection URL if we need to diff'},
    kilnFile: {alias: 'k', required: true, describe: 'path to file exporting kiln instance'},
  })
  .demandCommand(1).argv

function getUmzug() {
  const sequelize = getConnection({connectionURL: options.url})
  return new Umzug({
    logging: console.log,
    storage: 'sequelize',
    storageOptions: {sequelize},
    migrations: {params: [sequelize.getQueryInterface(), Sequelize]},
  })
}

async function migrate() {
  console.log('Starting migrations....')
  await getUmzug().up()
  console.log(colors.green('Migration successful'))
}

async function migrateRevert() {
  console.log('Starting reverse migrations....')
  await getUmzug().down()
  console.log(colors.green('Migration revert successful'))
}

async function migrationBootstrap(connectionURL) {
  const kilnFileExports = require(path.resolve(process.cwd(), options.kilnFile))
  const kiln = kilnFileExports.kiln || kilnFileExports
  if (typeof kiln.getModels !== 'function') {
    throw new TypeError('Did not export a kiln')
  }

  const connection = connectionURL && getConnection({connectionURL})
  const fileContent = await createNewMigrationFile(kiln, connection)
  const datePart = new Date()
    .toLocaleString(undefined, {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
    })
    .replace(/[^\d]+/g, '')

  const fileName = `migrations/${datePart}-migration.js`
  fs.writeFileSync(fileName, fileContent)
  console.log(colors.green('Wrote migration file '), fileName)
}

function run() {
  switch (options._[0]) {
    case 'migrate':
      return migrate()
    case 'migrate:revert':
      return migrateRevert()
    case 'migration:bootstrap':
      return migrationBootstrap()
    case 'migration:incremental':
      return migrationBootstrap(options.url)
    default:
      throw new Error(`Unrecognized command: ${options._[0]}`)
  }
}

run()
  .then(() => {
    console.log('🏁   Klay SQL Run Complete')
    process.exit(0)
  })
  .catch(err => {
    console.error(colors.red('💣   Fatal Error: \n'), err)
    process.exit(1)
  })
