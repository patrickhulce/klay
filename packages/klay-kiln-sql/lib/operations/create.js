const Promise = require('bluebird')
const validateAndAutomanageFactory = require('klay-db/helpers').validateAndAutomanage

const utils = require('../shared')

module.exports = function (modelDef, sequelizeModel, dependencies) {
  const validateAndAutomanage = validateAndAutomanageFactory(modelDef.model, 'create')
  const validateUnique = utils.validateUniqueConstraints(modelDef.model, dependencies)
  const validateCustom = utils.validateCustomConstraints(modelDef.model, dependencies)

  const operation = {
    _insert(object, options) {
      return sequelizeModel
        .create(utils.toStorage(modelDef.model, object), options)
        .then(record => utils.fromStorage(modelDef.model, record.get()))
    },
    run(object, options) {
      return Promise.resolve(object)
        .then(object => validateAndAutomanage(object, {failLoudly: true}).value)
        .then(object => validateUnique(object, null, options))
        .then(object => validateCustom(object, null, options))
        .then(object => operation._insert(object, options))
    },
  }

  return operation
}
