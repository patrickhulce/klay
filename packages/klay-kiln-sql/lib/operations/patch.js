var assert = require('assert');
var _ = require('lodash');
var Promise = require('bluebird');
var validateAndAutomanageFactory = require('klay-db/helpers').validateAndAutomanage;

var utils = require('../shared');
var updateOperation = require('./update');

module.exports = function (modelDef, sequelizeModel, dependencies) {
  var findByPrimaryKey = utils.findByPrimaryKey(modelDef.model, dependencies);
  var validateAndAutomanage = validateAndAutomanageFactory(modelDef.model, 'update');

  var update = updateOperation(modelDef, sequelizeModel, dependencies).run;

  var operation = {
    _mergeObject: function (patch, record) {
      assert.equal(typeof patch, 'object', 'patch must be an object');
      var merged = _.cloneDeep(record);
      _.forEach(patch, (v, k) => _.set(merged, k, v));
      return merged;
    },
    _update: function (patch, record, options) {
      var merged = operation._mergeObject(patch, record);
      return update(merged, options);
    },
    run: function (patch, options) {
      return Promise.resolve(patch).
        then(object => validateAndAutomanage(object).value).
        then(object => findByPrimaryKey(object, options)).
        then(record => operation._update(patch, record, options));
    },
  };

  return operation;
};
