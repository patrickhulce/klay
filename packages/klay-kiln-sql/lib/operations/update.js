var Promise = require('bluebird');
var validateAndAutomanageFactory = require('klay-db/helpers').validateAndAutomanage;

var utils = require('../shared');

module.exports = function (modelDef, sequelizeModel, dependencies) {
  var validateAndAutomanage = validateAndAutomanageFactory(modelDef.model, 'update');
  var validatePrimary = utils.validatePrimaryConstraint(modelDef.model, dependencies);
  var validateImmutable = utils.validateImmutableConstraints(modelDef.model, dependencies);
  var validateUnique = utils.validateUniqueConstraints(modelDef.model, dependencies);
  var validateCustom = utils.validateCustomConstraints(modelDef.model, dependencies);

  var operation = {
    _validateConstraints: function (object) {
      return validatePrimary(object).then(function (record) {
        return Promise.resolve().
          then(() => validateImmutable(object, record)).
          then(() => validateUnique(object, record)).
          then(() => validateCustom(object, record));
      });
    },
    _update: function (object, options) {
      return sequelizeModel.
        build(utils.toStorage(modelDef.model, object), {isNewRecord: false}).
        save(options).
        then(() => object);
    },
    run: function (object, options) {
      return Promise.resolve(object).
        then(object => validateAndAutomanage(object, {failLoudly: true}).value).
        then(object => operation._validateConstraints(object, options)).
        then(object => operation._update(object, options));
    },
  };

  return operation;
};
