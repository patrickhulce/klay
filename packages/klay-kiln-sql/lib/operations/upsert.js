var _ = require('lodash');
var utils = require('../shared');
var createOperation = require('./create');
var updateOperation = require('./update');

module.exports = function (modelDef, sequelizeModel) {
  var getPrimaryKey = utils.getPrimaryKey(modelDef.model);
  var findByPrimaryKey = utils.findByPrimaryKey(modelDef.model, sequelizeModel);
  var lookupRecords = utils.lookupByUniqueConstrains(modelDef.model, sequelizeModel);

  var create = createOperation(modelDef, sequelizeModel).run;
  var update = updateOperation(modelDef, sequelizeModel).run;

  var operation = {
    _lookupExistingRecord: function (object) {
      return lookupRecords(object).then(function (lookups) {
        if (!_.find(lookups, 'record')) {
          return;
        }

        var firstLookup = lookups[0];
        lookups.reduce(function (previous, current) {
          if (_.isEqual(previous.record, current.record)) {
            return current;
          } else {
            throw new Error('ambiguous upsert matches multiple records');
          }
        }, firstLookup);

        return findByPrimaryKey(firstLookup.record).then(function (existing) {
          return _.assign({}, existing, object);
        });
      });
    },
    run: function (object, options) {
      var validationResults = modelDef.model.validate(object);
      var primaryKey = getPrimaryKey(validationResults.value);

      if (primaryKey) {
        return update(object, options);
      } else {
        return operation._lookupExistingRecord(object).then(function (merged) {
          return merged ? update(merged, options) : create(object, options);
        });
      }
    },
  };

  return operation;
};
