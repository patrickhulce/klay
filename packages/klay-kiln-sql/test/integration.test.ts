describe('integration tests', () => {
  require('./integration/cli.test')
  require('./integration/initialize.test')
  require('./integration/create.test')
  require('./integration/custom-constraints.test')
  require('./integration/patch.test')
  require('./integration/query.test')
  require('./integration/transaction.test')
  require('./integration/update.test')
  require('./integration/upsert.test')
})
