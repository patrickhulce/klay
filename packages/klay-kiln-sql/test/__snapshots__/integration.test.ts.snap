// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration tests CLI migrate should execute migrations 1`] = `
Object {
  "age": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "BIGINT(20)",
  },
  "createdAt": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DATETIME(6)",
  },
  "email": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "VARCHAR(250)",
  },
  "firstName": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "VARCHAR(100)",
  },
  "id": Object {
    "allowNull": false,
    "defaultValue": null,
    "primaryKey": true,
    "type": "BIGINT(20)",
  },
  "isAdmin": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "TINYINT(1)",
  },
  "lastName": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "VARCHAR(100)",
  },
  "password": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "VARCHAR(32)",
  },
  "updatedAt": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DATETIME(6)",
  },
}
`;

exports[`integration tests CLI migrate should execute migrations 2`] = `
Object {
  "aspectRatio": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DOUBLE",
  },
  "createdAt": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DATETIME(6)",
  },
  "id": Object {
    "allowNull": false,
    "defaultValue": null,
    "primaryKey": true,
    "type": "CHAR(36)",
  },
  "metadata": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "TEXT",
  },
  "ownerId": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "BIGINT(20)",
  },
  "updatedAt": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DATETIME(6)",
  },
}
`;

exports[`integration tests CLI migrate should execute migrations 3`] = `
Array [
  "PRIMARY",
  "users_unique_email",
  "users_unique_firstname_lastname",
  "users_unique_firstname_lastname",
  "users_email_asc__password_asc",
  "users_email_asc__password_asc",
]
`;

exports[`integration tests CLI migration:bootstrap should have created a migration file 1`] = `
"module.exports = {
  up: async (queryInterface, Sequelize) => {
    await queryInterface.createTable('users', {
      id: {
        primaryKey: true,
        allowNull: false,
        autoIncrement: true,
        type: Sequelize.BIGINT
      },
      age: Sequelize.BIGINT,
      isAdmin: Sequelize.BOOLEAN,
      email: Sequelize.STRING(250),
      password: Sequelize.STRING(32),
      firstName: Sequelize.STRING(100),
      lastName: Sequelize.STRING(100),
      createdAt: Sequelize.DATE(6),
      updatedAt: Sequelize.DATE(6)
      })
    await queryInterface.addIndex('users', {'name':'users_unique_email','unique':true,'fields':['email']})
    await queryInterface.addIndex('users', {'name':'users_unique_firstname_lastname','unique':true,'fields':['firstName','lastName']})
    await queryInterface.addIndex('users', {'name':'users_email_asc__password_asc','fields':[{'attribute':'email','order':'ASC'},{'attribute':'password','order':'ASC'}]})
    await queryInterface.createTable('photos', {
      id: {
        primaryKey: true,
        allowNull: false,
        type: Sequelize.UUID
      },
      ownerId: {
        references: {
          key: 'id',
          model: 'users'
        },
        type: Sequelize.BIGINT
      },
      aspectRatio: Sequelize.DOUBLE,
      metadata: Sequelize.TEXT,
      createdAt: Sequelize.DATE(6),
      updatedAt: Sequelize.DATE(6)
      })
  },
  down: async (queryInterface, Sequelize) => {
    await queryInterface.dropTable('photos')
    await queryInterface.removeIndex('users', 'users_email_asc__password_asc')
    await queryInterface.removeIndex('users', 'users_unique_firstname_lastname')
    await queryInterface.removeIndex('users', 'users_unique_email')
    await queryInterface.dropTable('users')
  },
}"
`;

exports[`integration tests initialize database photos should have created a photos table 1`] = `
Object {
  "aspectRatio": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DOUBLE",
  },
  "createdAt": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DATETIME(6)",
  },
  "id": Object {
    "allowNull": false,
    "defaultValue": null,
    "primaryKey": true,
    "type": "CHAR(36)",
  },
  "metadata": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "TEXT",
  },
  "ownerId": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "BIGINT(20)",
  },
  "updatedAt": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DATETIME(6)",
  },
}
`;

exports[`integration tests initialize database users should have created a users table 1`] = `
Object {
  "age": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "BIGINT(20)",
  },
  "createdAt": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DATETIME(6)",
  },
  "email": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "VARCHAR(250)",
  },
  "firstName": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "VARCHAR(100)",
  },
  "id": Object {
    "allowNull": false,
    "defaultValue": null,
    "primaryKey": true,
    "type": "BIGINT(20)",
  },
  "isAdmin": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "TINYINT(1)",
  },
  "lastName": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "VARCHAR(100)",
  },
  "password": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "VARCHAR(32)",
  },
  "updatedAt": Object {
    "allowNull": true,
    "defaultValue": null,
    "primaryKey": false,
    "type": "DATETIME(6)",
  },
}
`;
